### DISCUSSION TITLE
MA27 solver

### DISCUSSION THREAD

**Discussion by Haugst3d** [coin-or/Sonnet#40] [General]
Hi
I'm trying to implement a MA27 solver in IpOpt using C#. I have the source building, and I'm thinking I should extend the OsiSolverInterface and ie an OsiMa27SolverInterface. But then I was thinking this would require me to wrap IpOpt library, and this must have been done before?

**Comment by jhmgoossens** [comment]
Sorry, I'm not sure I understand. How is your question related to Sonnet? 
But let me try to give some background, maybe that helps.

Sonnet can use any solver that supports the OsiSolverInterface. To be precise, [Sonnet contains .NET wrappers](https://github.com/coin-or/Sonnet/blob/master/src/SonnetWrapper/OsiClpSolverInterface.h) for some of the [native C++ OsiSolverInterface implementations](https://github.com/coin-or/Osi/tree/master/src). Sonnet itself calls these .NET wrappers, which in turn call the actual underlying native C++ base objects.

Out of the box, Sonnet is released with .NET wrappers for OsiCbc and OsiClp, but there is no big reason why any of the other Osi implementations could not be wrapped as well, like OsiCpx, etc. However, indeed there does not seem to be something like a native C++ OsiIpoptSolverInterface implementation or a OsiMA27SolverInterface. To create a .NET wrapper, you would first need to create the C++ version and then create a .NET wrapper for it. 

To implement a .NET wrapper for an Osi implementation, you _don't_ need to create .NET wrappers for the underlying solver C++ classes--the underlying native C++ libraries of the underlying solver are called from within the C++ base objects.

Does this help?

**Reply by Haugst3d** [reply]
Hi. It does help. I was inquiring in regards to the expandability of Sonnet. I like the structure and simplicity of Sonnet, but also require IpOpt for running a MA27 solver using Sonnet (from a .net app). But it looks like I will have to use a precompiled IpOpt as it does follow the same conventions with Visual Studio projects etc. so I cannot simply pull and link the ipopt repository in Sonnet and make a wrapper for the parts of ipopt I need. Is this correctly understood?

**Reply by jhmgoossens** [reply]
I haven't worked with ipopt, nor with MA27, so apologies if I get the details wrong:

>I was inquiring in regards to the expandability of Sonnet.
>[..] I will have to use a precompiled IpOpt as it does follow the same conventions with Visual Studio projects etc. 
>so I cannot simply pull and link the ipopt repository in Sonnet and make a wrapper for the parts of ipopt I need.

There are a couple of steps in this, and I can help with some of the later ones.

1. You'll need to implement something like C++ OsiIpOptSolverInterface, or OsiMA27SolverInterface or whatever makes sense. 
I suggest you to ask the IpOpt folks for guidance and help if this is conceptually even possible!
Honestly, I expect that if an OsiSolverInterface implementation for IpOpt is possible, it would have been created already..
Still, if that works out, then it would be great if you would contribute that Osi implementation back to the [coin-or/osi project](https://github.com/coin-or/osi).

2. You'll need windows binaries for IpOpt with MA27. 
This could indeed be the binaries released by IpOpt, but this would have to include the MA27 build.
But you can compile this also yourself. Indeed, IpOpt does not include the Visual Studio project & solution files, but it is built on Windows platforms. So, indeed, you cannot just include the projects in the Sonnet VS project, as is done for Cbc etc in Sonnet. But that's not a problem to do this with binaries, as long as these are DLLs and MSVS libs.

3. Sonnet needs a .NET wrapper around the new C++ OsiSolverInterface, and to call the binaries for IpOpt / MA27.
Making such a .NET wrapper once the C++ OsiSolverInterface is available is simple. If you can get OsiSolverInterface implementation added to the coin-or/Osi project then it should be simple to include this with a release of Sonnet.

As an alternative, you could also use Sonnet to create your problem instances, then export those into a file (MPS, LP, etc.) and import them into IpOpt and solve.
